core/vm/analysis.go

```diff
@@ -16,6 +16,102 @@
 
 package vm
 
+import (
+	"github.com/holiman/uint256"
+
+	"github.com/ledgerwatch/turbo-geth/common"
+)
+
+// fill in segment of operation information array for a block
+func analyzeBlock(ctx *callCtx, pc uint64) (*BlockInfo, error) {
+	blockInfo := NewBlockInfo(pc)
+	ctx.contract.opsInfo[pc] = blockInfo
+	code := ctx.contract.Code
+	codeLen := len(code)
+	jumpTable := ctx.interpreter.jt
+
+	height := 0
+	minHeight := 0
+	maxHeight := 0
+	for ; pc < uint64(codeLen); pc++ {
+		op := OpCode(code[pc])
+		oper := jumpTable[op]
+		if oper == nil {
+			continue
+		}
+
+		// track low and high watermark relative to block entry
+		height -= oper.numPop
+		minHeight = min(minHeight, height)	// will be <= 0
+		height += oper.numPush
+		maxHeight = max(maxHeight, height)	// will be >= 0
+		blockInfo.constantGas += oper.constantGas
+
+		if PUSH1 <= op && op <= PUSH32 {
+		    pushByteSize := int(op) - int(PUSH1) + 1
+
+			startMin := int(pc + 1)
+			if startMin >= codeLen {
+				startMin = codeLen
+			}
+			endMin := startMin + pushByteSize
+			if startMin+pushByteSize >= codeLen {
+				endMin = codeLen
+			}
+
+			integer := new(uint256.Int)
+			integer.SetBytes(common.RightPadBytes(
+				// So it doesn't matter what we push onto the stack.
+				code[startMin:endMin], pushByteSize))
+
+			// attach PushInfo with decoded push data to PUSHn
+			pushInfo := NewPushInfo(pc, *integer)
+			ctx.contract.opsInfo[pc] = pushInfo
+
+			continue
+		}
+		
+		// check jump destinations and optimize static jumps
+		if op == JUMP || op == JUMPI {
+			prevPC := pc - 1
+			prevOp := OpCode(code[prevPC])
```

Comment:
This looks both dangerous and wrong. 
1. The `pc -1` might eval to `-1`, causing a `panic` when accessing the `code[prevPC]`.
2. If the previous instruction is a `PUSH01`, then the previous PC was `pc -2`. If it indeed was a push instruction, then at `pc-1` there will always be pushdata. I think what you want is `ctx.contract.opsInfo[ len(ctx.contract.opsInfo) -1]` or something like that.

Reply:
I've been tweaking these and others, and they are closer to right.  JUMPDEST and JUMPI are subtle because they both end and begin blocks, and static jumps modify the code to improve performance.